File: sidecar/src/lume_sidecar/protocol.py
"""The Python side of the Rust <-> Python wire contract (DESIGN §9, §19).

This mirrors `crates/ipc/src/protocol.rs`. The same rule applies on both sides:
**nothing about model / device / framework crosses this seam** — only paths out,
and compressed vectors + JPEG thumbnails back. Raw pixels never cross (§6).

`to_dict` / `from_dict` use serde's default *externally-tagged* representation
for the `UnitResult` union (``{"Ok": {...}}`` / ``{"Failed": {...}}``) so the two
sides stay wire-compatible by construction. The concrete framing (length-prefixed
JSON header + binary payload vs. msgpack) is finalized when the transport lands
(BUILD.md M1); until then the round-trip tests pin the shape.
"""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class RequestUnit:
    """One Unit to embed. ``frame_ts`` None = whole image, else a video frame."""

    unit_idx: int
    path: str
    frame_ts: float | None = None

    def to_dict(self) -> dict:
        d: dict = {"unit_idx": self.unit_idx, "path": self.path}
        if self.frame_ts is not None:
            d["frame_ts"] = self.frame_ts
        return d

    @classmethod
    def from_dict(cls, d: dict) -> RequestUnit:
        return cls(unit_idx=d["unit_idx"], path=d["path"], frame_ts=d.get("frame_ts"))


@dataclass
class EmbedRequest:
    batch_id: int
    #: Stored grid-thumbnail edge, px (`Config.thumbnails.grid_px`). A
    #: legitimate batch-level parameter, not a model/device/framework leak.
    thumb_px: int
    units: list[RequestUnit] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {
            "batch_id": self.batch_id,
            "thumb_px": self.thumb_px,
            "units": [u.to_dict() for u in self.units],
        }

    @classmethod
    def from_dict(cls, d: dict) -> EmbedRequest:
        return cls(
            batch_id=d["batch_id"],
            thumb_px=d["thumb_px"],
            units=[RequestUnit.from_dict(u) for u in d["units"]],
        )


@dataclass
class UnitOk:
    """A successful embed: the fp16 vector bytes and the grid thumbnail JPEG."""

    emb_fp16: bytes
    thumb_jpeg: bytes


@dataclass
class UnitFailed:
    """An in-band per-unit failure (corrupt/unsupported file) — DESIGN §17."""

    reason: str


UnitResult = UnitOk | UnitFailed


def _unit_result_to_dict(r: UnitResult) -> dict:
    if isinstance(r, UnitOk):
        return {"Ok": {"emb_fp16": list(r.emb_fp16), "thumb_jpeg": list(r.thumb_jpeg)}}
    return {"Failed": {"reason": r.reason}}


def _unit_result_from_dict(d: dict) -> UnitResult:
    if "Ok" in d:
        ok = d["Ok"]
        return UnitOk(emb_fp16=bytes(ok["emb_fp16"]), thumb_jpeg=bytes(ok["thumb_jpeg"]))
    return UnitFailed(reason=d["Failed"]["reason"])


@dataclass
class BatchItem:
    unit_idx: int
    result: UnitResult

    def to_dict(self) -> dict:
        return {"unit_idx": self.unit_idx, "result": _unit_result_to_dict(self.result)}

    @classmethod
    def from_dict(cls, d: dict) -> BatchItem:
        return cls(unit_idx=d["unit_idx"], result=_unit_result_from_dict(d["result"]))


@dataclass
class EmbedResponse:
    batch_id: int
    items: list[BatchItem] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {"batch_id": self.batch_id, "items": [i.to_dict() for i in self.items]}

    @classmethod
    def from_dict(cls, d: dict) -> EmbedResponse:
        return cls(
            batch_id=d["batch_id"],
            items=[BatchItem.from_dict(i) for i in d["items"]],
        )


@dataclass
class EmbedOneRequest:
    """Synchronous drag-in image query. Response is one bare embedding vector."""

    image_bytes: bytes

    def to_dict(self) -> dict:
        return {"image_bytes": list(self.image_bytes)}

    @classmethod
    def from_dict(cls, d: dict) -> EmbedOneRequest:
        return cls(image_bytes=bytes(d["image_bytes"]))


@dataclass
class EmbedOneResponse:
    """One interactive query vector, shared by image and text query requests."""

    emb_fp16: bytes

    def to_dict(self) -> dict:
        return {"emb_fp16": list(self.emb_fp16)}

    @classmethod
    def from_dict(cls, d: dict) -> EmbedOneResponse:
        return cls(emb_fp16=bytes(d["emb_fp16"]))


@dataclass
class EmbedTextRequest:
    """Synchronous text-query embed for semantic search."""

    text: str

    def to_dict(self) -> dict:
        return {"text": self.text}

    @classmethod
    def from_dict(cls, d: dict) -> EmbedTextRequest:
        return cls(text=d["text"])

File: crates/ipc/src/protocol.rs
//! The Rust ↔ Python wire contract (DESIGN §9, §19 — BUILD.md L2).
//!
//! **Lock this early, never break it casually.** This is the load-bearing seam:
//! nothing about model / device / framework may ever appear in these messages.
//! If you find yourself adding a `device` or `model` field, the seam has leaked
//! — stop (DESIGN §19).
//!
//! Only *compressed/small* data crosses: a few KB of paths out, ~48 KB of fp16
//! vectors + ~1.3 MB of JPEG thumbnails back per 32-image batch. This is *why* a
//! plain Unix socket suffices and shared memory is rejected (§9). Raw pixels
//! never cross — the sidecar owns all decode (§6).
//!
//! These structs define the message *shape*. The concrete framing
//! (length-prefixed; JSON header + binary payload vs. msgpack) is finalized when
//! the transport lands — see `TODO(M1)` in `lib.rs`. Until then the round-trip
//! test pins the shape.

use serde::{Deserialize, Serialize};

/// Base64 ⇄ `Vec<u8>` serde adapter — the *one* place binary payload encoding
/// lives on the Rust side of the seam (issue #17, M2 transport depth).
///
/// DTO fields keep their `Vec<u8>` type so callers still see ordinary bytes,
/// but they cross the wire as compact base64 strings instead of JSON integer
/// arrays. Framing stays length-prefixed JSON, so frames remain inspectable.
mod b64 {
    use base64::engine::general_purpose::STANDARD;
    use base64::Engine as _;
    use serde::{Deserialize, Deserializer, Serializer};

    pub fn serialize<S: Serializer>(bytes: &[u8], serializer: S) -> Result<S::Ok, S::Error> {
        serializer.serialize_str(&STANDARD.encode(bytes))
    }

    pub fn deserialize<'de, D: Deserializer<'de>>(deserializer: D) -> Result<Vec<u8>, D::Error> {
        let encoded = String::deserialize(deserializer)?;
        STANDARD
            .decode(encoded.as_bytes())
            .map_err(serde::de::Error::custom)
    }
}

/// One request frame from Rust to Python.
///
/// The tag is the socket-level operation name. Keeping this envelope explicit
/// lets the Sidecar server multiplex bulk indexing, drag-in image queries, and
/// text queries on one length-prefixed transport without leaking model details.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", content = "payload", rename_all = "snake_case")]
pub enum ClientMessage {
    Embed(EmbedRequest),
    EmbedOne(EmbedOneRequest),
    EmbedText(EmbedTextRequest),
}

/// One response frame from Python to Rust.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", content = "payload", rename_all = "snake_case")]
pub enum ServerMessage {
    EmbedResponse(EmbedResponse),
    EmbedOneResponse(EmbedOneResponse),
    Error { message: String },
}

/// One Unit to embed in a batch request.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct RequestUnit {
    /// Index within this batch — the response echoes it so results re-align even
    /// if the sidecar reorders internally.
    pub unit_idx: u32,
    pub path: String,
    /// `None` = whole image; `Some(ts)` = video frame at `ts` seconds.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub frame_ts: Option<f32>,
}

/// Rust → Python: embed a batch (DESIGN §9, bulk path).
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct EmbedRequest {
    pub batch_id: u64,
    /// Stored grid-thumbnail edge, px (`Config.thumbnails.grid_px`). A
    /// legitimate batch-level parameter, not a model/device/framework leak
    /// (DESIGN §19) — the sidecar still owns all decode/resize knowledge.
    pub thumb_px: u32,
    pub units: Vec<RequestUnit>,
}

/// One Unit's result. Per-unit failure is in-band, not fatal (DESIGN §17).
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub enum UnitResult {
    Ok {
        /// 768 little-endian fp16 elements = 1536 bytes. Base64 on the wire.
        #[serde(with = "b64")]
        emb_fp16: Vec<u8>,
        /// Grid thumbnail, JPEG-encoded (DESIGN §8). Base64 on the wire.
        #[serde(with = "b64")]
        thumb_jpeg: Vec<u8>,
    },
    Failed {
        reason: String,
    },
}

/// Python → Rust: results for a batch, echoing `batch_id`.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct EmbedResponse {
    pub batch_id: u64,
    pub items: Vec<BatchItem>,
}

/// A response item, carrying its `unit_idx` for re-alignment.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct BatchItem {
    pub unit_idx: u32,
    pub result: UnitResult,
}

/// Rust → Python: synchronous single-image embed for drag-in queries (§12, M4).
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct EmbedOneRequest {
    #[serde(with = "b64")]
    pub image_bytes: Vec<u8>,
}

/// Python → Rust: the drag-in query vector.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct EmbedOneResponse {
    #[serde(with = "b64")]
    pub emb_fp16: Vec<u8>,
}

/// Rust → Python: synchronous text-query embed for semantic search (§12).
///
/// The response reuses [`EmbedOneResponse`]: both interactive paths return one
/// bare query vector and no thumbnail.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct EmbedTextRequest {
    pub text: String,
}
